### PR TITLE
bake: fix printer handling

### LINKED ISSUE #2598: bake: panic in progress printer
https://github.com/docker/buildx/actions/runs/9915811338/job/27397269631?pr=2595#step:6:259

```
  panic: runtime error: invalid memory address or nil pointer dereference
  [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x13d7ba7]
  
  goroutine 10 [running]:
  github.com/docker/buildx/util/progress.(*Printer).Warnings(...)
  	github.com/docker/buildx/util/progress/printer.go:65
  github.com/docker/buildx/commands.runBake.func2()
  	github.com/docker/buildx/commands/bake.go:149 +0x27
  github.com/docker/buildx/util/progress.NewPrinter.func1()
  	github.com/docker/buildx/util/progress/printer.go:129 +0x20a
  created by github.com/docker/buildx/util/progress.NewPrinter in goroutine 1
  	github.com/docker/buildx/util/progress/printer.go:112 +0x2b1
```

https://github.com/docker/buildx/blob/10c9ff901c1c255b58969dd2ade4789acbbab3bb/util/progress/printer.go#L65

### PR DESCRIPTION
**Pull request by crazy-max**
relates to #2598

Removes extra go routine that is racy when used along the other one from `progress.WithOnClose`.

### CODE REVIEW

File: commands/bake.go [hunk 1]
```diff
@@ -238,26 +207,23 @@ func runBake(ctx context.Context, dockerCli command.Cli, targets []string, in ba
 		return err
 	}
 
-	def := struct {
+	dtdef, err := json.MarshalIndent(struct {
```
Comment (tonistiigi) [inline commands/bake.go:210]: Passing this encoded JSON around seems weird. I think we can just call encode in SaveLocalState.

**Review by tonistiigi — COMMENTED** [review]
I think this mostly is good but if we want to cherry-pick then I think for v0.16 we should go with something like

```
if p := printer; p != nil {
```

**Comment by tonistiigi** [comment]
@crazy-max yes

File: commands/bake.go [hunk 2]
```diff
@@ -468,12 +424,34 @@ func bakeCmd(dockerCli command.Cli, rootOpts *rootOptions) *cobra.Command {
...
 }
 
-func saveLocalStateGroup(dockerCli command.Cli, ref string, lsg localstate.StateGroup) error {
+func saveLocalStateGroup(dockerCli command.Cli, in bakeOptions, targets []string, bo map[string]build.Options, overrides []string, def any) error {
+	prm := confutil.MetadataProvenance()
+	if len(in.metadataFile) == 0 {
+		prm = confutil.MetadataProvenanceModeDisabled
+	}
+	groupRef := identity.NewID()
+	var refs []string
```
Comment (tonistiigi) [inline commands/bake.go:433]: nit: `refs := make([]string, 0, len(bo))`

**Review by tonistiigi — APPROVED** [review]

File: commands/bake.go (hunk 2 above)
Comment (crazy-max) [reply commands/bake.go]: fixed

**Pull request merged**